File: 0013-roman-to-integer/0013-roman-to-integer.cpp
class Solution {
public:
    int romanToInt(string s) {
        int total = 0;
        int n = s.size();
        
        auto value = [](char c) {
            switch(c) {
                case 'I': return 1;
                case 'V': return 5;
                case 'X': return 10;
                case 'L': return 50;
                case 'C': return 100;
                case 'D': return 500;
                case 'M': return 1000;
            }
            return 0;
        };
        
        for (int i = 0; i < n; i++) {
            if (i + 1 < n && value(s[i]) < value(s[i + 1])) {
                total += value(s[i + 1]) - value(s[i]);
                i++; 
            } else {
                total += value(s[i]);
            }
        }
        
        return total;
        unordered_map<char, int> mp = {{'I', 1},   {'V', 5},   {'X', 10},
                                       {'L', 50},  {'C', 100}, {'D', 500},
                                       {'M', 1000}};
        int n = s.size();
        int number = 0;
        for (int i = 0; i < n; i++) {
            if (i + 1 < n && mp[s[i]] < mp[s[i + 1]]) {
                number -= mp[s[i]];
            } else {
                number += mp[s[i]];
            }
        }
        return number;

    }
};

File: 0013-roman-to-integer/0013-roman-to-integer.java
class Solution {
    public int romanToInt(String s) {
        int ans=0;
        for(int i=0;i<s.length();i++){
            if(s.charAt(i)=='M'){
                if(i>0 && s.charAt(i-1)=='C' ){
                    ans+=800;
                }
                else ans+=1000;
            }
            else if(s.charAt(i)=='D'){
                if(i>0 && s.charAt(i-1)=='C' ){
                    ans+=300;
                }
                else ans+=500;
            }
            else if(s.charAt(i)=='C'){
                if(i>0 && s.charAt(i-1)=='X' ){
                    ans+=80;
                }
                else ans+=100;
            }
            else if(s.charAt(i)=='L'){
                if(i>0 && s.charAt(i-1)=='X' ){
                    ans+=30;
                }
                else ans+=50;
            }
            else if(s.charAt(i)=='X'){
                if(i>0 && s.charAt(i-1)=='I' ){
                    ans+=8;
                }
                else ans+=10;
            }
            else if(s.charAt(i)=='V'){
                 if(i>0 && s.charAt(i-1)=='I' ){
                    ans+=3;
                }
                else ans+=5;
            }
            else if(s.charAt(i)=='I'){
                ans+=1;
            }
        }
        return ans;
    }
}